tests/Microsoft.Identity.Test.Unit.net45/pop/PopAuthenticationSchemeTests.cs

```diff
@@ -0,0 +1,97 @@
+        }
+
+        [TestMethod]
+        public void ValidatePopRequestAndToken()
+        {
+            // Arrange
+            Uri uri = new Uri("https://www.contoso.com/path1/path2?queryParam1=a&queryParam2=b");
+            HttpMethod method = HttpMethod.Post;
+            var popCryptoProvider = Substitute.For<IPoPCryptoProvider>();
+            popCryptoProvider.CannonicalPublicKeyJwk.Returns(JWK);
+            const string AtSecret = "secret";
+            MsalAccessTokenCacheItem msalAccessTokenCacheItem = new MsalAccessTokenCacheItem()
+            {
+                Secret = AtSecret
+            };
+
+            // Act
+            PoPAuthenticationScheme authenticationScheme = new PoPAuthenticationScheme(uri, method, popCryptoProvider);
+            var tokenParams = authenticationScheme.GetTokenRequestParams();
+            var popTokenString = authenticationScheme.FormatAccessToken(msalAccessTokenCacheItem);
+            JwtSecurityToken decodedPopToken = new JwtSecurityToken(popTokenString);
+
+            // Assert
+            Assert.AreEqual("POP", authenticationScheme.AuthorizationHeaderPrefix);
+            Assert.AreEqual(JWT, authenticationScheme.KeyId);
+            Assert.AreEqual(2, tokenParams.Count);
+            Assert.AreEqual("pop", tokenParams["token_type"]);
+
+            // This is the base64 URL encoding of the JWK containing only the KeyId
+            Assert.AreEqual("eyJraWQiOiJOemJMc1hoOHVEQ2NkLTZNTndYRjRXXzdub1dYRlpBZkhreFpzUkdDOVhzIn0", tokenParams["req_cnf"]);
+            Assert.AreEqual("RS256", decodedPopToken.Header.Alg);
+            Assert.AreEqual(JWT, decodedPopToken.Header.Kid);
+            Assert.AreEqual("pop", decodedPopToken.Header.Typ);
+            Assert.AreEqual("RS256", decodedPopToken.SignatureAlgorithm);
+
+            AssertSimpleClaim(decodedPopToken, "at", AtSecret);
+            AssertSimpleClaim(decodedPopToken, "m", HttpMethod.Post.ToString());
+            AssertSimpleClaim(decodedPopToken, "u", "www.contoso.com");
+            AssertSimpleClaim(decodedPopToken, "p", "/path1/path2");
+
+            string nonce = AssertSimpleClaim(decodedPopToken, "nonce");
+            Assert.IsFalse(string.IsNullOrEmpty(nonce));
+            string jwk = AssertSimpleClaim(decodedPopToken, "cnf");
+            var jwkFromPopAssertion = JToken.Parse(jwk);
+            
+            var initialJwk = JToken.Parse(JWK);
+            Assert.IsTrue(jwkFromPopAssertion["jwk"].DeepEquals(initialJwk));
+        }
+
+        private static string AssertSimpleClaim(JwtSecurityToken jwt, string expectedKey, string optionalExpectedValue = null)
+        {
+            string value = jwt.Claims.Single(c => c.Type.Equals(expectedKey, StringComparison.InvariantCultureIgnoreCase)).Value;
+            if (optionalExpectedValue != null)
+            {
+                Assert.AreEqual(optionalExpectedValue, value);
+            }
+            return value;
+        }
+    }
```

Comment:
do you have any tests for catching the ests errors for failed requests? spec says ESTS will respond with Bad Request if the req_cnf parameter is provided when token_type=pop and the input value is one of the following:
Not URL encoded
Not containing a kid field
Containing more elements than kid

is it possible to have tests for that? do we do anything special in msal for these cases? also, this may be more middleware related...still wrapping my head around this. :)

Reply:
This is a very good question. In this PR, we do not allow developers to specify their own key, because on Windows we can do that for them. But it is a natural extensibility point which we will undertake sooner or later. 

Because we do key management, I don't expect ESTS to ever return these errors, because MSAL obeys the rules above. However, implementers other than MSAL (e.g. if you use call the Web APIs directly), then you might hit those if you do not package the key correctly.

Reply:
makes sense. thanks.

---
In reply to: [352715068](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1525#discussion_r352715068) [](ancestors = 352715068)